clang_build/compiler.py

```diff
@@ -135,26 +134,156 @@ def _dialect_exists(self, year, clangpp):
+            if "draft" in line or "gnu" in line:
+                continue
+
+            return "-std=" + _search(r"'(c\+\+..)'", line).group(1)
+
+        raise RuntimeError("Could not find a supported C++ standard.")
+
+    def _get_driver(self, source_file):
+        if source_file.suffix in [".c", ".cc", ".m"]:
+            return [str(self.clang)]
+        else:
+            return [str(self.clangpp), self.max_cpp_dialect]
+
+    def compile(self, source_file, object_file, flags):
+        """Compile a given source file into an object file.
+
+        If the object file is placed into a non-existing folder, this
+        folder is generated before compilation.
+
         Parameters
         ----------
-        clangpp : :any:`pathlib.Path`
-            Path to the clang++ executable
+        source_file : pathlib.Path
+            The source file to compile
+
+        object_file : pathlib.Path
+            The object file to generate during compilation
+
+        flags : list of str
+            List of flags to pass to the compiler
 
         Returns
         -------
+        bool
+            True if the compilation was successful, else False
         str
-            Flag string of the latest supported dialect
+            Output of the compiler
 
         """
-        supported_dialects = []
-        for dialect in range(30):
-            if self._dialect_exists(dialect, clangpp):
-                supported_dialects.append(dialect)
+        object_file.parents[0].mkdir(parents=True, exist_ok=True)
 
-        if supported_dialects:
-            return self._get_dialect_flag(max(supported_dialects))
-        else:
-            return self._get_dialect_flag(98)
+        return self._run_clang_command(
+            self._get_driver(source_file)
+            + ["-c", str(source_file), "-o", str(object_file)]
+            + flags
+        )
+
+    def generate_dependency_file(self, source_file, dependency_file, flags):
+        """Generate a dependency file for a given source file.
+
+        If the dependency file is placed into a non-existing folder, this
+        folder is generated before compilation.
+
+        Parameters
+        ----------
+        source_file : pathlib.Path
+            The source file to compile
+
+        dependency_file : pathlib.Path
+            The dependency file to generate
+
+        flags : list of str
+            List of flags to pass to the compiler
+
+        Returns
+        -------
+        bool
+            True if the dependency file generation was successful, else False
+        str
+            Output of the compiler
+
+        """
+        dependency_file.parents[0].mkdir(parents=True, exist_ok=True)
+
+        return self._run_clang_command(
+            self._get_driver(source_file)
+            + ["-E", "-MMD", str(source_file), "-MF", str(dependency_file)]
+            + flags
+        )
+
+    def link(self, output_file, command, output_type):
```

Comment:
Linking is part of clang's feature. We could decide if we do not call it `compiler.py` but placing functionality part of clang's functionality in a different class seems odd to me.

Reply:
I can change the interface to the link function, though.

Reply:
From a terminology point of view, "flags" is also not 100% accurate for both compile and link as it also contains include directories etc., which I guess you wouldn't typically call "flags".

Reply:
I think the static "archiver" rather than linker also has issues. I think the static part is what made it more difficult for me to actually try to come up with a good solution as the current implementation as far as I can tell is arbitrary and not necessarily what I for example would want from a static library:

1) It adds object files of other dependencies. For me it is unclear why this is always wanted.
2) It does not add object files of other dependencies _recursively_. This might also not be what people would expect.

Reply:
I believe `clang` and `clangpp`, when used for linking, are two drivers for `lld`, but `llvm-ar` is not `clang` in my opinion (in the gcc case, it's `ld`).
However, I agree it sort of belongs together - but I cannot think of a fitting name to encompass compiler + linker...

Reply:
No llvm-ar on linux is ar. It doesn't link, it only archives. I will try to come up with a better name.

Reply:
> From a terminology point of view, "flags" is also not 100% accurate for both compile and link as it also contains include directories etc., which I guess you wouldn't typically call "flags".

Indeed. It might be best to create the `-I /path/to/include` commands in the `Compiler` class and add `include_directories` and `library_directories` to these functions.
By the way it bothered me a little that the former is a possible parameter in the project config, but the latter is not.

> No llvm-ar on linux is ar. It doesn't link, it only archives.

Right. The definitions are (correct me if I'm wrong)
- object files are created by a compiler
- static libraries are created by an archiver
- dynamic libraries / shared objects are created by a linker
- executables are created by a linker

"builder" would not be a great name IMO and "assembler" is already taken ^^

Reply:
We could call it `toolsets.py` and inside you have the `Clang` class. And maybe at some point also the `GCC` class etc. I agree with the include_directories etc. This would also help further abstract away compiler specific terminology such as "-I".

Reply:
I suggest `toolchain.py` and calling the class `LLVM`.